### PR TITLE
Simplification of the core struct and related methods

### PR DESCRIPTION
**Pull request by RaunoArike**
To simplify the core struct and related methods, I propose the following changes:

- Create new structs for all different convergence criteria, and dispatch based on the types of these structs when calling the `converged()` method. This removes a ton of kwargs from the main struct and makes the code clearer in general. For the more casual users, nothing changes after this refactor: they can still use the `:converge_when` keyword argument to specify a convergence criterion with the default arguments. The more advanced user will need to be a bit more involved than before to define custom convergence criteria: they will have to define one of the new convergence structs on their own and pass that to the `generate_counterfactual()` method. However, the improved clarity of the code seems to justify the added hassle for those users.
- Remove the `converged()` method that had been created specifically for Growing Spheres. There's no reason Growing Spheres shouldn't fit into the pre-existing convergence framework.
- Remove the `:early_stopping` convergence type, which appears to be functionally the same as the `:max_iter` convergence type.
- Move the `generative_model_params` kwarg out from the `CounterfactualExplanation` struct and define it as a kwarg of the `GradientBasedGenerator` struct instead, since it's only related to the generators.
- Remove the `learning_rate` keyword argument, as it's not used anywhere in the package.
- Move the files that were previously in the `generate_counterfactual` folder to the `counterfactuals` folder. Since the primary functionality of the methods in that folder is instantiating `CounterfactualExplanation` structs, which are defined in the `counterfactuals` folder, it makes more sense for these methods to be defined inside the same folder with the struct.
- This is a slightly unrelated change, but the `Generator` struct was not used anywhere in the package, so I also removed that, as well as a few other pieces of redundant code.

Also, sorry for the enormous size of the PR. I'll keep the next ones shorter.

### CODE REVIEW

File: src/generators/gradient_based/loss.jl [hunk 1]
```diff
@@ -39,9 +37,15 @@ function ∇(
...
+
+"""
+    hinge_loss(convergence::InvalidationRateConvergenc, ce::AbstractCounterfactualExplanation)
+
+The default hinge loss for any convergence criterion.
+Can be overwritten inside the `convergence` folder as part of the definition of specific convergence criteria.
+"""
+function hinge_loss(convergence::AbstractConvergence, ce::AbstractCounterfactualExplanation)
+    return 0
 end
```
Comment (RaunoArike) [inline src/generators/gradient_based/loss.jl:51]: I created this method because defining a default `hinge_loss()` method inside the `Generators` module and overriding it inside the `Convergence` module for the `InvalidationRateConvergence` criterion seemed to be the only way of avoiding cyclical dependencies between the `Convergence` and `Generators` modules. If I had defined `hinge_loss` only inside the `Convergence` module, I would have had to call `Convergence.hinge_loss()` inside the `∇` method, while if I had defined `hinge_loss` only inside the `Generators` module, I would have had to call `Convergence.InvalidationRateConvergence` inside the `hinge_loss()` method. The `Convergence` module has to make calls to the `Generators` module anyways, so there's no way to allow `Convergence` to be a dependency of `Generators`. I'm aware that this solution is not ideal, though, so if anyone has better ideas, let me know.

**Comment by RaunoArike** [comment]
@pat-alt @VincentPikand @kmariuszk given the size of the PR, I think it's best if all of you take at least a quick look at the PR and let me know whether you agree with the changes, so marked all of you as reviewers

**Review by pat-alt — COMMENTED** [review]
Overall this looks really good! Very clean and understandable. Please see my smaller comments below. I still need to look at some of the core files more closely and will then approve (later today or tomorrow morning). 

Regarding the comment on moving `hinge_loss` and the other one to `Objectives`, we can just open a new issue for that. 

Great work!

File: src/generators/non_gradient_based/growing_spheres/generate_perturbations.jl
[the remark this answers is not in the document]
Comment (RaunoArike) [reply src/generators/non_gradient_based/growing_spheres/generate_perturbations.jl:1]: I'll delete this file, it was included in the PR by accident: it was created as part of PR #344 which I merged into this one at some point, but we ended up deleting that file inside #344. I'll check to make sure no other such unintentional changes are included in this PR.

File: src/convergence/invalidation_rate.jl
[the remark this answers is not in the document]
Comment (RaunoArike) [reply src/convergence/invalidation_rate.jl:1]: On a second thought, defining those methods in the `Objectives` module indeed seems like the best option. However, this would still create a circular dependency: `Convergence` would depend on `Generators`, `Generators` would depend on `Objectives`, and `Objectives` would have to depend on `Convergence` to enable dispatching on the convergence type. I'll try to find a solution to this later today.

File: src/convergence/max_iter.jl
[the remark this answers is not in the document]
Comment (RaunoArike) [reply src/convergence/max_iter.jl:1]: I modified the `converged()` method, does this version look better?

[the remark this answers is not in the document]
Comment (pat-alt) [reply src/convergence/max_iter.jl:1]: Sorry, actually I think the previous version was correct, as long as there is still also a `terminated` condition (can't easily tell from here without checking this out locally). Basically, in case the search is terminated because `max_iter` has been reached, we the `terminated` condition to be true but the `converged` condition false. Unless the threshold is reached in the final iteration I guess.

[the remark this answers is not in the document]
Comment (RaunoArike) [reply src/convergence/max_iter.jl:1]: There is still a `terminated` condition for all convergence criteria, yep. If we want `converged` to be true if the threshold is reached in the final iteration, then I guess `MaxIterationConvergence` should function exactly like `DecisionThresholdConvergence`, except for the fact that `MaxIterationConvergence` would check whether the threshold has been reached only after the final iteration, while `DecisionThresholdConvergence` would check it after every iteration?

[the remark this answers is not in the document]
Comment (pat-alt) [reply src/convergence/max_iter.jl:1]: OK, I've just had a look locally. I think the current version is the clearest, even though the previous version hat the exact same effect since 

```julia
function terminated(ce::CounterfactualExplanation)
    return Convergence.converged(ce.convergence, ce) || steps_exhausted(ce)
end
```

I've added a note to the doc string to make it clear that `MaxIterationConvergence` truly means: "converge when maximum number of iterations has been reached and ignore everything else (except timer)".

File: src/counterfactuals/core_struct.jl
Comment (pat-alt) [inline src/counterfactuals/core_struct.jl:1]: Gotta love how much cleaner this is now!

File: src/counterfactuals/initialisation.jl
Comment (pat-alt) [inline src/counterfactuals/initialisation.jl:1]: This now adds random perturbations even for counterfactuals using latent space search. That seems redundant but it also doesn't seem to hurt and is probably more consistent. So happy if you're all happy with this.

Also: Not in this PR, but for consistency should we treat initialization in the same way as convergence? What do you people think?

**Review by pat-alt — APPROVED** [review]

File: src/convergence/invalidation_rate.jl
[the remark this answers is not in the document]
Comment (VincentPikand) [reply src/convergence/invalidation_rate.jl:1]: After discussing this with Rauno, I believe that we should keep it here for now and open a separate PR. The circular dependency problem needs a bit more work to be elegant.

[the remark this answers is not in the document]
Comment (pat-alt) [reply src/convergence/invalidation_rate.jl:1]: Actually, it looks to me like you can simply define `hinge_loss(ce::AbstractCounterfactualExplanation)` just like other loss functions, because the convergence struct is part of `ce`?

**Review by pat-alt — APPROVED** [review]

File: src/generators/gradient_based/loss.jl [hunk 2]
```diff
@@ -39,9 +37,15 @@ function ∇(
     M::Models.AbstractDifferentiableModel,
     ce::AbstractCounterfactualExplanation,
 )
-    ℓ = 0
-    if (ce.convergence[:converge_when] == :invalidation_rate)
-        ℓ = hinge_loss(ce)
-    end
-    return ∂ℓ(generator, M, ce) + ∂h(generator, ce) .+ ℓ
+    return ∂ℓ(generator, M, ce) + ∂h(generator, ce) .+ hinge_loss(ce.convergence, ce)
```
Comment (pat-alt) [inline src/generators/gradient_based/loss.jl:40]: Actually I just stumbled across this. That can't be right? Now it looks like every gradient-based generator also get's hinge loss?

**Review by pat-alt — CHANGES_REQUESTED** [review]

File: src/generators/gradient_based/loss.jl (hunk 2 above)
Comment (pat-alt) [reply src/generators/gradient_based/loss.jl:40]: Oh I see, you overloaded that method to return 0 by default. Hmmm let me have a quick go at this before we merge please.

Comment (pat-alt) [reply src/generators/gradient_based/loss.jl:40]: Also doesn't look like Probe is now actually using these loss functions (correct me if I'm missing something)

```julia
function ProbeGenerator(;
    λ::AbstractFloat=0.1,
    loss::Symbol=:logitbinarycrossentropy,
    penalty::Penalty=Objectives.distance_l1,
    kwargs...,
)
    @assert haskey(losses_catalogue, loss) "Loss function not found in catalogue."
    user_loss = Objectives.losses_catalogue[loss]
    return GradientBasedGenerator(; loss=user_loss, penalty=penalty, λ=λ, kwargs...)
end
```

Comment (pat-alt) [reply src/generators/gradient_based/loss.jl:40]: Another thing that doesn't seem to check out above is that hinge loss isn't differentiated. We should improve tests here to catch such things.

**Review by VincentPikand — CHANGES_REQUESTED** [review]
I think some things still need to be sorted out before we can merge

File: src/counterfactuals/initialisation.jl [hunk 3]
```diff
@@ -8,19 +8,7 @@ Initializes the starting point for the factual(s):
...
-
-    # No perturbation:
-    if ce.initialization == :identity
-        return s′
-    end
-
-    # If latent space, initial point is random anyway:
-    if ce.params[:latent_space]
-        return s′
-    end
```
[the remark this answers is not in the document]
Comment (RaunoArike) [reply src/counterfactuals/initialisation.jl:23]: The function returns unmodified `s′` by default, so we don't need if-statements for the cases where we want to return `s′`, we only need if-statements for the cases where we want to return something else.

File: src/convergence/Convergence.jl [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+Returns the convergence object from the dictionary of default convergence types.
+"""
+function get_convergence_type(convergence::Symbol)
+    return get(
+        convergence_catalogue,
+        convergence,
+        () -> error("Convergence criterion not recognized: $convergence."),
+    )
+end
```
Comment (kmariuszk) [inline src/convergence/Convergence.jl:47]: image

Using this `get` method you introduce incompatibility with Julia 1.6 and lower. Is it necessary to use it?

File: src/generators/gradient_based/loss.jl (hunk 2 above)
Comment (RaunoArike) [reply src/generators/gradient_based/loss.jl]: > Also doesn't look like Probe is now actually using these loss functions (correct me if I'm missing something)

This is correct, the user has to specify `InvalidationRateConvergence` as the convergence type to use Probe in the intended way. It's the same way it was before: though `hinge_loss()` was defined in the same folder as `ProbeGenerator`, Probe actually called the `hinge_loss()` method only if the user defined `converge_when = :invalidation_rate` inside the core struct (see the deleted line 43 in `src/generators/gradient_based/loss.jl`). The reason I'm dispatching on `hinge_loss()` based on the convergence type rather than based on whether the user defined Probe as the generator is that I want the generators to remain composable: this way, you can also use `hinge_loss()` and `invalidation_rate` with other generators if you want.

File: src/convergence/Convergence.jl (hunk 4 above)
Comment (RaunoArike) [reply src/convergence/Convergence.jl:47]: We're using neither tuples nor numbers here, so I guess it's fine?

File: src/counterfactuals/initialisation.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (RaunoArike) [reply src/counterfactuals/initialisation.jl:23]: Wait, now I noticed that `:latent_space` can be combined with the `:add_perturbations` initialization type. Since it doesn't matter for `:latent_space` whether perturbations are added or not, though, it still seems more consistent to always just do that if initialization is set to `:add_perturbations`.

File: src/counterfactuals/core_struct.jl [hunk 5]
```diff
@@ -28,12 +27,7 @@ end
...
 		num_counterfactuals::Int = 1,
 		initialization::Symbol = :add_perturbation,
-		generative_model_params::NamedTuple = (;),
-		min_success_rate::AbstractFloat=0.99,
-        converge_when::Symbol=:decision_threshold,
-        invalidation_rate::AbstractFloat=0.5,
-        learning_rate::AbstractFloat=1.0,
-        variance::AbstractFloat=0.01,
+        convergence::Union{AbstractConvergence,Symbol}=:decision_threshold,
 	)
```
[the remark this answers is not in the document]
Comment (RaunoArike) [reply src/counterfactuals/core_struct.jl]: Fixed

**Review by kmariuszk — APPROVED** [review]

**Comment by VincentPikand** [comment]
> For some reason I can't reply directly to this comment:
> 
> > It is differentiated, see lines 39
> 
> That seems wrong to me: looking at the paper, it seems we first need taking gradients to compute the approximation of IR (which is what's done here). But then we still need to differentiate the counterfactual search with respect to that estimated quantity.

I don't remember enough to answer this currently and as such we can make this a separate issue as it is kind of unrelated to the changes of this PR.

**Comment by pat-alt** [comment]
Hi team, I tried to fix this thinking it would be quick work:

> > For some reason I can't reply directly to this comment:
> > > It is differentiated, see lines 39
> > 
> > 
> > That seems wrong to me: looking at the paper, it seems we first need taking gradients to compute the approximation of IR (which is what's done here). But then we still need to differentiate the counterfactual search with respect to that estimated quantity.
> 
> I don't remember enough to answer this currently and as such we can make this a separate issue as it is kind of unrelated to the changes of this PR.

But @VincentPikand was right, should have parked this for a new issue and PR (sorry!). I've reverted everything to 0f2560e here now and opened #371 which contains all the modifications that I've made, so let's take it from there. There's another inconsistency I've noted and will detail in that draft PR. 

There are a few commits here such as [b77e8a8](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/pull/360/commits/b77e8a8559861f2bb569cce87e5a3cc3c57565f8) and [e6fcaa0](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/pull/360/commits/e6fcaa0a5967028ad879bd2bc2bc75dc96206b20) that I've reverted that we might want to re-revert (sorry again!). 

Other than that, happy to merge this one

**Comment by VincentPikand** [comment]
@pat-alt (and my future employers who are looking into my open source contributions), pardon my french, but this PR is fucked up beyond repair. We've agreed to split this up as much as possible into digestible bites instead of 23-1 fasting megatron meals.
